content/en/blog/_posts/2021-06-08-Kubeadm-Avoid-Default-Address.md

```diff
@@ -0,0 +1,285 @@
+because each node is a VirtualBox VM that uses what VirtualBox calls a
+"NAT network adapter" to get connectivity to the outside world.  You
+may have other use cases, perhaps involving containers.
+
+In this post I will look at how to make all of the changes needed for
+a single-control-node cluster.  I have not yet attempted multiple
+control nodes, and the solution will need further work for that case.
+The needed results are getting the desired address to appear in the
+following places.
+
+- The IP address that the kube-apiserver puts in its Endoints object.
+- The Subject Alternative Name list of the x509 certificate generated
+  for the kube-apiserver.
+- The IP address that appears in the `kubeadm join` commands that
+  `kubeadm` outputs.
+- The InternalIP type of address that each kubelet puts into its
+  Node's status.
+
+This post describes work done with version 1.21.1 of kubeadm and
+kubernetes.
+
+You may also like to read a related blog post, [Kubernetes Setup Using
+Ansible and
+Vagrant](https://kubernetes.io/blog/2019/03/15/kubernetes-setup-using-ansible-and-vagrant/).
+In contrast, the post at hand is about comprehensively overriding the
+default address and not being specific to one form of automation.
+
+As an added bonus, I will also mention how to stop DHCP from assigning
+the same address to every node for networks where that is not desired.
+This comes first because everything else depends on it; you can skip
+that section if you are not having that problem.
+
+## Getting DHCP to assign unique addresses
+
+This is a problem that I had when using a common Ubuntu 20.04 snapshot
+to base my various VMs on.  This snapshot includes a file named
+`/etc/machine-id`, and so all my VMs get the same content in that file
+(unless I take extra steps).  This is relevant because in Ubuntu
+20.04, the default DHCP client is `systemd.networkd`, and its default
+configuration causes it to send an explicit "DHCP client identifier"
+(in addition to the client MAC address) in each request and that
+identifier is derived from `/etc/machine-id` and _not_ the relevant
+network interface's MAC address.  When an explicit client ID is
+supplied, the DHCP server uses only that --- not also the client MAC
+address.  Howevever, rather than having to tweak each derived VM, I
+tweaked the common snapshot to make the DHCP client eschew that
+explicit client ID --- leaving the DHCP server to look only at the
+client's MAC address.  The relevant bit of `systemd.networkd`
+configuration is normally driven by a higher level tool called
+`netplan`.  I tweaked its configuration by creating a drop-in file
+named `/etc/netplan/10-enp0s8.yaml` (because `enp0s8` is the name of
+the network interface that I want to get a distinct IPv4 address on
+each node) with the following contents.
+
+```yaml
+network:
+  ethernets:
+    enp0s8:
+      dhcp4: true
+      dhcp-identifier: mac
+  version: 2
+```
+
+## The kube-apiserver's address in its Endpoints object
+
```

Comment:
Maybe remind readers what an Endpoints is (I think the kube-apiserver creates a Service in the default namespace, and I'm not 100% sure on how the related Endpoints / EndpointSlices come into being) - it's a niche topic for some readers.

Reply:
Added.